streamer/src/bidirectional_channel.rs

```diff
@@ -0,0 +1,396 @@
+                .insert(current_id, sender_channel.clone());
+            data.sender_socket_address_map
+                .insert(quic_address.clone(), current_id);
+            // create a new or replace the exisiting id
+            data.sender_ids_map
+                .insert(current_id, sender_channel.clone());
+            data.sender_socket_address_map
+                .insert(quic_address.clone(), current_id);
+            current_id
+        } else {
+            0
+        };
+
+        if sender_id == 0 {
+            return;
+        }
+
+        let subscriptions = self.data.clone();
+        // start listnening to stream specific cross beam channel
+        tokio::spawn(async move {
+            tokio::pin!(send_stream);
+            loop {
+                let reciever_channel = reciever_channel.clone();
+                let recv_task = async move { reciever_channel.recv() };
+                let send_stream_close_task = async { send_stream.stopped() };
+                tokio::select! {
+                    task = recv_task => {
+                        match task {
+                            Ok(message) => {
+                                let serialized_message = serialize(&message).unwrap();
+                                let _ =send_stream.write(serialized_message.as_slice()).await;
+                            },
+                            Err(_) => {
+                                // disconnect
+                                let _ = send_stream.finish().await;
+                                break;
+                            }
+                        }
+                    },
+                    _task = send_stream_close_task => {
+                        // disconnect
+                        let _ = send_stream.finish().await;
+                        break;
+                    }
+                };
+            }
+
+            // remove all data belonging to sender_id
+            let write_lock = subscriptions.write();
+            if let Ok(mut sub_data) = write_lock {
+                let subscriptions = &mut *sub_data;
+                subscriptions.sender_ids_map.remove(&sender_id);
+                subscriptions
+                    .transaction_signature_map
+                    .retain(|_, v| *v != sender_id);
+                subscriptions
+                    .sender_socket_address_map
+                    .retain(|_, v| *v != sender_id);
+            }
+        });
+    }
+
+    pub fn add_packets(&self, quic_address: &SocketAddr, packets: &PacketBatch) {
+        // check if socket is registered;
+        let id = {
+            let data = self.data.read();
+            if let Ok(data) = data {
+                data.sender_socket_address_map
+                    .get(quic_address)
+                    .map_or(0, |x| *x)
+            } else {
+                0
+            }
+        };
+        if id == 0 {
+            return;
+        }
+        let data = self.data.write();
```

Comment:
i think it's better to upgrade the existing read lock instead of locking, unlocking and relocking